app/helpers/glossary_helper.rb

```diff
@@ -0,0 +1,85 @@
+module GlossaryHelper
+  DEFAULT_OPTIONS = {once: true}.freeze
+
+  def link_glossary_terms(doc, options = DEFAULT_OPTIONS)
+    @_glossary_pattern ||=
+      begin
+        @_glossary = {}
+
+        names = []
+        GlossaryTerm.all.each do |term|
+          term.names.map(&:downcase).each do |name|
+            @_glossary[name] = term
+          end
+
+          names.concat(term.names)
+        end
+
+        names.sort.reverse.map do |name|
+          Regexp.escape(name.downcase)
+        end.join("|")
+      end
+
+    doc.traverse do |child|
+      next unless child.name == "text" && glossary_context?(child)
+
+      found, es = replace_glossary_terms(
+        doc, child.content,
+        @_glossary.dup, @_glossary_pattern
+      )
+      child.swap(Nokogiri::XML::NodeSet.new(doc.document, es))
+
+      if options[:once]
+        @_glossary.delete_if{ |_, term| found.include?(term) }
+      end
+    end
+  end
+
+  private
+
+  def glossary_context?(context)
+    required_context = ["p", "li"]
+    excluded_contexts = ["a"]
+
+    context.ancestors.map do |node|
+      return false if excluded_contexts.include?(node.name)
+
+      required_context.include?(node.name)
+    end.any?
+  end
+
+
+  def replace_glossary_terms(doc, text, map, pattern)
+    found, elements = [], []
+    sample = text
+
+    while match = sample.match(/\b(#{pattern})\b/i)
+      break if match[0].empty?
+
+      term = map.delete(match[0].downcase)
+
+      if term
+        found << term
+
+        link = Nokogiri::XML::Element.new("a", doc)
```

Comment:
Is it necessary to create these tags in the helper? I think it's more idiomatic (and, therefore, discoverable) to create an object, like a list of GlossaryTerms or a Struct that responds to the necessary methods, and then pass it to a view partial.

Reply:
In this case we're not generating much HTML but instead doing some transformations on the HTML spit out by our wysiwyg. Nokogiri provides a very nice and safe way to do that.

The markup for each glossary linked segment could be put in a partial with something like

```erb
   <%= segment.pre_term %><%= link_to segment.term, class: "glossary-term" do %>
      <%= segment.term.name %> <%= image_path("info.png") %>
   <% end %><%= segment.post_term %>
```

But it would awkward and wasteful then to render that, just to parse it again with Nokogiri so that nodes can be inserted into it's model of the document. It's better IMO to just use the api Nokogiri gives us.